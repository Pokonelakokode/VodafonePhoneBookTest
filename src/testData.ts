export const testUser = [{name: "TEST",phone:"1234678",email:"test@email.com",id:1}];