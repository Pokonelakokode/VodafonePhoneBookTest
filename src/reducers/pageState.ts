import {IPageStateActions, PageStateActionTypes} from "../actions/pageState";
import {getHash} from "../utils";

export type PageState = {
    selectedUser: number | null,
    search: string,
    letterFilter: string
}

const initialState:PageState = {
    selectedUser: parseInt(getHash().user) || null,
    search: "",
    letterFilter: ""
};


export default function pageState(state = initialState, action: IPageStateActions) {
    switch (action.type) {
        case PageStateActionTypes.SET:
            return {...state,[action.key]:action.data};
        default:
            return state
    }
}

