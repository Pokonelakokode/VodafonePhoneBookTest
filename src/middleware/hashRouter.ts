import {Dispatch, Middleware} from "redux";
import {RootActions} from "../actions";
import {PageStateActionTypes} from "../actions/pageState";
import {setHash} from "../utils";

export const hashRouter:Middleware = () => (next:Dispatch) => (action:RootActions) => {
    next(action);
    if(action.type === PageStateActionTypes.SET && action.key === "selectedUser"){
        action.data ? setHash({user:action.data}) : setHash();
    }
};