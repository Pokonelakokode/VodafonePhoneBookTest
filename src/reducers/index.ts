import { combineReducers } from "redux";
import pageState from "./pageState";
import users from "./users";

export const rootReducer = combineReducers({
    pageState,
    users
});

export type RootState = ReturnType<typeof rootReducer>