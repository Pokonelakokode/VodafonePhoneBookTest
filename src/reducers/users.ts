import {IUserActions, UserActionTypes} from "../actions/users";
import { User } from "../types";
import {loadFromStorage} from "../utils";

export default function users(state = loadFromStorage(), action: IUserActions):User[] {
    switch (action.type) {
        case UserActionTypes.LOAD:
            return action.users;
        case UserActionTypes.SET:
            return state.map(user => user.id === action.user.id ? action.user : user);
        case UserActionTypes.ADD:
            return [...state,action.user];
        case UserActionTypes.DELETE:
            return state.filter(user => user.id !== action.id);
        default:
            return state
    }
}