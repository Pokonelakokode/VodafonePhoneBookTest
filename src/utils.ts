import  users from './users.json';
import { User } from './types.js';

export function saveToStorage(data:User[] = []) {
    window.localStorage.setItem("users",JSON.stringify(data));
    return data
}

export function validateEmail(email:string):boolean {
    var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(String(email).toLowerCase());
}

export function loadFromStorage():User[] {
    let data = window.localStorage.getItem("users");
    if(data){
        try {
            return JSON.parse(data);
        }
        catch (e) {
            saveToStorage(users);
            return users
        }
    }
    else {
        saveToStorage(users);
        return users
    }
}

export function getHash() {
    let hash = window.location.hash.substr(1);
    return hash.split('&').reduce((acc:{[key:string]:any},el) => {
        const data = el.split('=');
        if(data.length === 2) acc[data[0]] = data[1];
        return acc
    },{})
}

export function setHash(data: { [key: string]: any } = {},push=true) {
    data = isObject(data) ? data : {};
    const hash = Object.keys(data).reduce((acc:string,el:any) => {
        acc += `${el}=${data[el]}&`;
        return acc;
    },"/#").slice(0,-1);
    push ? history.pushState({user:data.user || null},"",hash) : history.replaceState({user: data.user},"",hash)
}

export function isObject (value:any) {
    return value && typeof value === 'object' && value.constructor === Object;
}
