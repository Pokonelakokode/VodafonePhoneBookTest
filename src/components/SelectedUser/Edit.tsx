import * as React from "react";
import {useEffect, useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {mainSelector} from "../../actions";
import {addUser, deleteUser, getUserOrNew, saveUser} from "../../actions/users";
import {bindActionCreators} from "redux";
import {selectUser} from "../../actions/pageState";
import {validateEmail} from "../../utils";

interface IProps {
    selectedUser: number
}


const Edit: React.FC<IProps> = ({selectedUser}) => {
    if(selectedUser === null) return <h1>No such user</h1>;
    const {users} = useSelector(mainSelector);
    const actions = bindActionCreators({saveUser, addUser, selectUser, deleteUser}, useDispatch());

    const [user, setUser] = useState(getUserOrNew(users, selectedUser)[0]);
    const [edit, setEdit] = useState(!user.id);
    useEffect(() => {
        let [user, isNew] = getUserOrNew(users, selectedUser);
        setUser(user);
        setEdit(isNew);
    }, [selectedUser]);
    const editCancel = () => {
        if (edit) setUser(getUserOrNew(users, selectedUser)[0]);
        setEdit(!edit);
    };
    const _deleteUser = () => {
        if (confirm(`Are you sure you want to delete ${user.name} from contacts?`)) {
            actions.deleteUser(user.id);
            actions.selectUser(null)
        }
    };
    const nameValid = user.name.length > 0;
    const emailValid = validateEmail(user.email);
    const phoneValid = user.phone.length >= 7;
    const _saveUser = () => {
        let isNew: boolean;
        if(!nameValid || !emailValid || !phoneValid) return alert("Some of your data is invalid");
        [user.id, isNew] = !user.id ? [users.reduce((max: number, _user) => _user.id > max ? _user.id : max, -1) + 1, true] : [user.id, false];
        if (isNew) {
            actions.addUser(user);
            actions.selectUser(user.id);
        } else {
            actions.saveUser(user);
        }
        setEdit(false);
    };

    return (
        <div>
            <h1 className="text-center">{!user.id ? "Add New Contact" : `${edit ? "Edit " : ""}${user.name}'s Profile`}</h1>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Name</label>
                <div className="col-sm-10">
                    <input value={user.name}
                           placeholder="Enter Name"
                           readOnly={!edit} type="text"
                           className={`${nameValid ? "" : "is-invalid "}form-control`}
                           required
                           onChange={(e) => setUser({...user, name: e.target.value})}
                    />
                    {!nameValid && <div className="invalid-feedback d-block">Name is too short</div>}
                </div>

            </div>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Email</label>
                <div className="col-sm-10">
                    <input value={user.email}
                           placeholder="Enter Email"
                           readOnly={!edit} type="email"
                           className={`${emailValid ? "" : "is-invalid "}form-control`}
                           onChange={(e) => setUser({...user, email: e.target.value})}
                    />
                    {!emailValid && <div className="invalid-feedback d-block">Email address is invalid</div>}
                </div>
            </div>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Phone</label>
                <div className="col-sm-10">
                    <input value={user.phone} placeholder="Enter Phone"
                           readOnly={!edit} type="text"
                           required
                           className={`${phoneValid ? "" : "is-invalid "}form-control`}
                           onChange={(e) => setUser({...user, phone: e.target.value})}
                    />
                    {!phoneValid && <div className="invalid-feedback d-block">Phone number have to be at least 7 digits</div>}
                </div>
            </div>
            <div className="row justify-content-around">
                {user.id &&
                <button onClick={editCancel} className="btn btn-primary">{edit ? "Cancel" : "Edit"}</button>}
                <button onClick={() => actions.selectUser(null)} className="btn btn-primary">Back</button>
                {edit ?
                    <button onClick={_saveUser} className="btn btn-danger">{user.id ? "Save" : "Create"}</button>
                    :
                    user.id && <button onClick={_deleteUser} className="btn btn-danger">Delete</button>
                }
            </div>

        </div>
    )
};

export default Edit