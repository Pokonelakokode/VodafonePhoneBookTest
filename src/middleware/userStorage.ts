import {Dispatch, Middleware, MiddlewareAPI} from "redux";
import {RootActions} from "../actions";
import {UserActionTypes} from "../actions/users";
import {saveToStorage} from "../utils";
import {RootState} from "../reducers";

export const userStorage:Middleware = (store: MiddlewareAPI<any,RootState>) => (next:Dispatch) => (action:RootActions) => {
    const {DELETE, ADD, SET} = UserActionTypes;
    next(action);
    if(action.type === DELETE || ADD || SET){
        saveToStorage(store.getState().users)
    }
};

export default userStorage