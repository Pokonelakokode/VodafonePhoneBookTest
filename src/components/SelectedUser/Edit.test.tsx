import * as React from "react";
import * as renderer from "react-test-renderer"
import {createStore, applyMiddleware} from "redux";
import { Provider } from "react-redux";
import { rootReducer } from "../../reducers";
import { hashRouter } from "../../middleware/hashRouter";
import { userStorage } from "../../middleware/userStorage";
import Edit from "./Edit";
import { mount } from "enzyme";
import {saveToStorage} from "../../utils";
import {testUser} from "../../testData";

describe("Edit",() => {
    saveToStorage(testUser);
    const store = createStore(rootReducer, applyMiddleware(hashRouter, userStorage));

    it('should render snapshot', function () {
        const component = renderer.create(
            <Provider store={store}>
                <Edit selectedUser={1}/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    it('should be able to set edit and save', function () {
        const wrapper = mount(
            <Provider store={store}>
                <Edit selectedUser={1}/>
            </Provider>
        );
        wrapper
            .find('button')
            .at(0)
            .simulate('click');
        expect(wrapper.find('button').at(0).text()).toBe("Cancel");
        wrapper
            .find('input')
            .at(0)
            .simulate('change',{target: {value: ""}});
        expect(wrapper.find('input').at(0).text()).toBe("");
        expect(wrapper.find('div.invalid-feedback').at(0).text()).toBe("Name is too short");
        wrapper
            .find('input')
            .at(0)
            .simulate('change',{target: {value: "TEST2"}});
        wrapper
            .find('input')
            .at(1)
            .simulate('change',{target: {value: "TEST2@email.com"}});
        wrapper
            .find('input')
            .at(2)
            .simulate('change',{target: {value: "1234567"}});
        wrapper
            .find('button')
            .at(2)
            .simulate('click');
        expect(store.getState().users.find(user => user.id === 1)).toStrictEqual({id:1,name:"TEST2",email:"TEST2@email.com",phone:"1234567"})
    });
    it('should be able to create new User and delete it', function () {
        window.confirm = jest.fn(() => true);
        const wrapper = mount(
            <Provider store={store}>
                <Edit selectedUser={-1}/>
            </Provider>
        );
        expect(wrapper.find('button').at(0).text()).toBe("Back");
        expect(wrapper.find('button').at(1).text()).toBe("Create");
        wrapper
            .find('input')
            .at(0)
            .simulate('change',{target: {value: "TEST3"}});
        wrapper
            .find('input')
            .at(1)
            .simulate('change',{target: {value: "TEST3@email.com"}});
        wrapper
            .find('input')
            .at(2)
            .simulate('change',{target: {value: "1234567"}});
        wrapper
            .find('button')
            .at(1)
            .simulate('click');
        expect(store.getState().users.find(user => user.id === 2)).toStrictEqual({id:2,name:"TEST3",email:"TEST3@email.com",phone:"1234567"});
        wrapper
            .find('button')
            .at(2)
            .simulate('click');
        expect(window.confirm).toBeCalled();
        expect(store.getState().users.length).toBe(1);
    });
    it('should nullify the selectedUser when Back button clicked', function () {
        const wrapper = mount(
            <Provider store={store}>
                <Edit selectedUser={-1}/>
            </Provider>
        );
        wrapper
            .find('button')
            .at(0)
            .simulate('click');
        expect(store.getState().pageState.selectedUser).toBe(null);
    });
});