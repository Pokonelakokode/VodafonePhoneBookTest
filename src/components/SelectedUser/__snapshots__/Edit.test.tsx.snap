// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit should render snapshot 1`] = `
<div>
  <h1
    className="text-center"
  >
    TEST's Profile
  </h1>
  <div
    className="form-group row"
  >
    <label
      className="col-sm-2 col-form-label"
    >
      Name
    </label>
    <div
      className="col-sm-10"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Enter Name"
        readOnly={true}
        required={true}
        type="text"
        value="TEST"
      />
    </div>
  </div>
  <div
    className="form-group row"
  >
    <label
      className="col-sm-2 col-form-label"
    >
      Email
    </label>
    <div
      className="col-sm-10"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Enter Email"
        readOnly={true}
        type="email"
        value="test@email.com"
      />
    </div>
  </div>
  <div
    className="form-group row"
  >
    <label
      className="col-sm-2 col-form-label"
    >
      Phone
    </label>
    <div
      className="col-sm-10"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Enter Phone"
        readOnly={true}
        required={true}
        type="text"
        value="1234678"
      />
    </div>
  </div>
  <div
    className="row justify-content-around"
  >
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Back
    </button>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;
